File: 0525 contiguous-array/solution.go
func findMaxLength(nums []int) int {
    arr:=make([]int,2*len(nums)+1)
    for i,_ := range arr {
        arr[i]=-2
    }
    arr[len(nums)]=-1
    sum,ma:=len(nums),0

    for i,v:=range nums {
        sum+=v*2-1
        if arr[sum]!=-2 {
            ma=max(ma,i-arr[sum])
        } else {
            arr[sum]=i
        }
    }
    return ma
}

File: 0525 contiguous-array/solution.java
class Solution {
    public int findMaxLength(int[] nums) {
        int[] arr = new int[2*nums.length+1];
        Arrays.fill(arr,-2);
        arr[nums.length]=-1;
        int sum=nums.length,max=0;
        for(int i=0;i<nums.length;i++) {
            sum+=nums[i]*2-1;
            if(arr[sum]!=-2) {
                max=Math.max(max,i-arr[sum]);
            } else {
                arr[sum]=i;
            }
        }
        return max;
    }
}
